optuna/samplers/_qmc.py

```diff
@@ -0,0 +1,357 @@
+    def _log_incomplete_reseeding() -> None:
+        _logger.warning(
+            "The seed of QMC seqeunce is not reseeded and only the seed of `independent_sampler` "
+            "is reseeded. This is to ensure that each workers samples from the same QMC sequence "
+            "in the parallel and/or distributed environment."
+            "You can suppress this warning by setting `warn_reseeding` "
+            "to `False` in the constructor of `QMCSampler`, "
+            "if this random seeding is intended behavior."
+        )
+
+    def sample_independent(
+        self,
+        study: Study,
+        trial: FrozenTrial,
+        param_name: str,
+        param_distribution: BaseDistribution,
+    ) -> Any:
+
+        if self._initial_search_space is not None:
+            if self._warn_independent_sampling:
+                self._log_independent_sampling(trial, param_name)
+
+        return self._independent_sampler.sample_independent(
+            study, trial, param_name, param_distribution
+        )
+
+    def sample_relative(
+        self, study: Study, trial: FrozenTrial, search_space: Dict[str, BaseDistribution]
+    ) -> Dict[str, Any]:
+
+        if search_space == {}:
+            return {}
+
+        sample = self._sample_qmc(study, search_space)
+        trans = _SearchSpaceTransform(search_space)
+        sample = trans.bounds[:, 0] + sample * (trans.bounds[:, 1] - trans.bounds[:, 0])
+        return trans.untransform(sample[0, :])
+
+    def after_trial(
+        self,
+        study: "optuna.Study",
+        trial: "optuna.trial.FrozenTrial",
+        state: TrialState,
+        values: Optional[Sequence[float]],
+    ) -> None:
+        self._independent_sampler.after_trial(study, trial, state, values)
+
+    def _sample_qmc(
+        self, study: Study, search_space: Dict[str, BaseDistribution]
+    ) -> numpy.ndarray:
+
+        # Lazy import because the `scipy.stats.qmc` is slow to import.
+        import scipy.stats.qmc
+
+        sample_id = self._find_sample_id(study, search_space)
+        d = len(search_space)
+
+        # Use cached `qmc_engine` or construct a new one.
+        if self._is_engine_cached(d, sample_id):
+            qmc_engine: scipy.stats.qmc.QMCEngine = self._cached_qmc_engine
+        else:
+            if self._qmc_type == "halton":
+                qmc_engine = scipy.stats.qmc.Halton(d, seed=self._seed, scramble=self._scramble)
+            elif self._qmc_type == "sobol":
+                qmc_engine = scipy.stats.qmc.Sobol(d, seed=self._seed, scramble=self._scramble)
+            else:
+                message = (
+                    f"The `qmc_type`, {self._qmc_type}, is not a valid. "
+                    'It must be one of "halton" and "sobol".'
+                )
+                raise ValueError(message)
+
+        forward_size = sample_id - qmc_engine.num_generated  # `sample_id` starts from 0.
+        qmc_engine.fast_forward(forward_size)
+        sample = qmc_engine.random(1)
```

Comment:
I do not believe fast `forward(...)` and `random(1)` give warnings in this situation.

Reply:
It would only print a warning if `qmc_engine.num_generated==0` and if you call `qmc_engine.random` with something else than a power of 2.

If you never saw the warning here, it means you are fast forwarding on the first call the Sobol' sequence. If it's really the case, remember it's bad to skip the first point. You should not do any burnin with Sobol'. If you must, only burn 2^m points and then you can only sample 2^n with n<=m.

Reply:
I think `qmc_engine.random(1)` after the initialization of the engine does not raise a warning. For example, the following code should give you no warning.
```python
import scipy.stats
en = scipy.stats.qmc.Sobol(3)
en.random(1)
```

Reply:
Sorry my previous comment might be off the point for you. By the design of optuna, it is not possible to know the total number of samples in advance. Therefore it is in principle impossible to warn the user about the total size of samples here. 

As far as I can remember, that was a reason to use Halton sequence instead of Sobol by default. I think it is OK to just have a caveat on size of the Sobol sequence in the docstring (line 63).

Reply:
Indeed the current warning is skipped for the first point `(n & (n - 1) == 0)` is not working with `n = 0`. Not sure it needs a fix. We can argue a single point is a digital net. Anyway, fine here if you cannot warn the user. And warning at the end would not be optimal either except if the user could ask for n more points to match 2*m. But nobody would care I guess.